### PR TITLE
Support errors.tolerance

### PR DESCRIPTION
**Pull request by rozza**
Ensure that missing or invalid Resume Tokens do not error.

KAFKA-105

Scenarios tested:

*Source*
 
 * Missing / invalid / not found Resume Tokens (Integration test with a mocked offsetStorageReader) 
 * Poison pill message - invalid schema

*Sink*

  * Poison pill messages - Invalid Key / Values types & invalid documents
  * Errors thrown by PostProcessors
  * Debezium CDC handler errors / poison pills

 Evergreen: https://spruce.mongodb.com/version/5f620eba2fbabe32998f3c09/tasks

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@
   - [KAFKA-154](https://jira.mongodb.org/browse/KAFKA-154) Improve the handling and error messaging for Json array config values.
   - [KAFKA-78](https://jira.mongodb.org/browse/KAFKA-78) Added dead letter queue support for the source connector.
   - [KAFKA-157](https://jira.mongodb.org/browse/KAFKA-157) Improved error message for business key errors.
+  - [KAFKA-105](https://jira.mongodb.org/browse/KAFKA-105) Improve `errors.tolerance=all` support in the sink and source connectors.
```
Comment (rozza) [inline CHANGELOG.md:23]: `errors.tolerance=all` is a KAFKA connect setting which we've copied.  It allows errors to be ignored outside of the connector (eg. during conversion).  This PR adds support inside the connector where possible, config errors are still errors but message based errors during `Source.poll` & `Sink.put` now honor this setting.

File: src/integrationTest/java/com/mongodb/kafka/connect/sink/MongoSinkTaskIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+import org.junit.runner.RunWith;
+
+import org.bson.Document;
+
+import com.mongodb.client.MongoCollection;
+
+import com.mongodb.kafka.connect.mongodb.MongoKafkaTestCase;
+
+@RunWith(JUnitPlatform.class)
+public class MongoSinkTaskIntegrationTest extends MongoKafkaTestCase {
```
Comment (rozza) [inline src/integrationTest/java/com/mongodb/kafka/connect/sink/MongoSinkTaskIntegrationTest.java:51]: Normal "happy" case tested first then various error scenarios.

File: src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java [hunk 3]
```diff
@@ -319,6 +342,40 @@ void testSourceCanHandleNonExistentCollectionAndSurviveDropping() {
...
+              put(MongoSourceConfig.ERRORS_TOLERANCE_CONFIG, ErrorTolerance.ALL.value());
+              put(MongoSourceConfig.POLL_MAX_BATCH_SIZE_CONFIG, "50");
+              put(MongoSourceConfig.POLL_AWAIT_TIME_MS_CONFIG, "1000");
+            }
+          };
+
+      String offsetToken =
+          "{\"_data\": \"825F58DDF4000000032B022C0100296E5A1004BBCFDF90907247ABA61D94DF01D76200461E5F6964002B020004\"}";
+      when(context.offsetStorageReader()).thenReturn(offsetStorageReader);
+      when(offsetStorageReader.offset(any())).thenReturn(singletonMap("_id", offsetToken));
```
Comment (rozza) [inline src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java:368]: Here we use a mock to provide an invalid / not existing offsetToken

File: src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java [hunk 4]
```diff
@@ -493,6 +550,105 @@ void testDeadletterQueueHandling() {
...
+                  "{\"type\" : \"record\", \"name\" : \"fullDocument\","
+                      + "\"fields\" : [{\"name\": \"_id\", \"type\": \"int\"}]}");
+            }
+          };
+
+      task.start(cfg);
+
+      Document poisionPill = Document.parse("{_id: {a: 1, b: 2, c: 3}}");
+      insertMany(rangeClosed(1, 3), coll);
+      coll.insertOne(poisionPill);
```
Comment (rozza) [inline src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java]: The poison pill does not match the schema

File: src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java [hunk 5]
```diff
@@ -493,6 +550,105 @@ void testDeadletterQueueHandling() {
...
+                  "{\"type\" : \"record\", \"name\" : \"fullDocument\","
+                      + "\"fields\" : [{\"name\": \"_id\", \"type\": \"int\"}]}");
+            }
+          };
+
+      task.start(cfg);
+
+      Document poisonPill = Document.parse("{_id: {a: 1, b: 2, c: 3}}");
+      insertMany(rangeClosed(1, 3), coll);
+      coll.insertOne(poisonPill);
```
Comment (rozza) [inline src/integrationTest/java/com/mongodb/kafka/connect/source/MongoSourceTaskIntegrationTest.java:580]: The poison document does not match the schema

File: src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java [hunk 6]
```diff
@@ -234,12 +230,25 @@ private AtomicInteger getRemainingRetriesForTopic(final String topic) {
...
+        }
+      }
+      if (!config.tolerateErrors()) {
+        throw new DataException("Failed to write mongodb documents despite retrying", e);
+      }
+    } else {
+      Integer deferRetryMs = config.getInt(RETRIES_DEFER_TIMEOUT_CONFIG);
+      LOGGER.info("Deferring retry operation for {}ms", deferRetryMs);
+      context.timeout(deferRetryMs);
+      throw new RetriableException(e.getMessage(), e);
```
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java:250]: The RetriableException is Kafkas own exception handling mechanism.

File: src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java [hunk 7]
```diff
@@ -127,10 +148,25 @@ public BsonDocument clone() {
...
-        unwrapped = supplier.get();
-      } catch (Exception e) {
-        throw new DataException("Unexpected data conversion exception.", e);
+      switch (dataType) {
+        case KEY:
+          try {
+            unwrapped = converter.apply(record.keySchema(), record.key());
+          } catch (Exception e) {
+            throw new DataException(
+                format("Could not convert key `%s` into a BsonDocument.", record.key()), e);
```
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java:157]: By using a BiFunction the record can be accessed so meaningful error messages can be outputted.

File: src/main/java/com/mongodb/kafka/connect/sink/converter/SinkConverter.java [hunk 8]
```diff
@@ -43,24 +45,25 @@ public SinkDocument convert(final SinkRecord record) {
...
+              Type.VALUE,
+              (Schema schema, Object data) ->
+                  getRecordConverter(schema, data).convert(schema, data));
     }
 
     return new SinkDocument(keyDoc, valueDoc);
   }
 
-  private RecordConverter getRecordConverter(final Object data, final Schema schema) {
+  private RecordConverter getRecordConverter(final Schema schema, final Object data) {
```
Comment (rozza) [inline src/main/java/com/mongodb/kafka/connect/sink/converter/SinkConverter.java:66]: Consistent param ordering.

**Review by durran — APPROVED** [review]

**Pull request merged**